### DISCUSSION TITLE
Question about GetStartupInfo

### DISCUSSION THREAD

**Discussion by chazste** [PlummersSoftwareLLC/HelloAssembly#26] [Ideas]
@davepl 
    It looks to me that the call to GetStartupInfo is vestigial code. I think originally it was present to set up the nCommandShow parameter to WinMain(), which then gets passed to the ShowWindow() API call. When WinMain() was collapsed into MainEntry, the ShowWindow() call was replaced by UpdateWindow() so there is no longer any place for the wShowWindow parameter to go. It just gets pushed on the stack and then never used. 
    I believe removing that stanza should be OK, since the result is not used. It will also remove a few bytes from the IAT.

**Comment by davepl** [comment]
It still needs to respect the original startup style for the window, though.  Windows can launch an app minimized or maximized or regular, and I think that’s passed in the startup info.  So long as that’s not lost.

It sounds like a regression if that’s being ignored?  It doesn’t have to go through WinMain like it used to but still need to support the startup style.

- Dave

**Comment by chazste** [comment]
OK. My reading of the code says that this is a regression then. I will do some testing later and go from there.

**Reply by chazste** [reply]
@davepl 
    I tested Tiny.exe using a shortcut with the "Run" field set to each of the three available values. TinyOriginal never calls ShowWindow(). Nevertheless the docs for ShowWindow() describe what we do: we call CreateWindow() with the WS_VISIBLE flag set. In this case when the window is displayed the OS consults the STARTUPINFO structure to see how to display the main (first) window. If you agree that this test was valid, I would say the call to GetStartupInfo() is not needed and is simply left over from when there was a need to create the parameter list for WinMain().

**Comment by davepl** [comment]
As along as it still works, that’s all I care.  I added that to make Raymond Chen happy, so don’t want to remove it like it didn’t matter :-)

**Comment by chazste** [comment]
@davepl 
I believe this one has a high enough payback to continue this discussion. I am a big fan of @oldnewthing and have read his blog for many years. So I would like to not only make the result smaller, I'd like it to still be correct. (And not just work by accident)

I went back and read Raymond's blog post on [STARTF_USESHOWWINDOW](https://devblogs.microsoft.com/oldnewthing/20100301-00/?p=14773) as well as the MSDN entries on [ShowWindow](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-showwindow), [CreateWindowEx](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-createwindowexa) and [WinMain](https://learn.microsoft.com/en-us/windows/win32/learnwin32/winmain--the-application-entry-point)

My summary of all four sources is that the window manager will feel free to override your provided argument to ShowWindow on the first call for the application's main window. Since we don't care about how the window is shown, and only that it is shown as requested by the starting application we should be good without calling GetStartupInfo. Again we don't call ShowWindow directly, so even trying to control the second argument to ShowWindow is even more convoluted when going through CreateWindowEx.

**Comment by davepl** [comment]
Thanks for doing the research!  Are you saying, then, that the window manager has a hack to do it for you because most apps don’t do it right?  If so, then we don’t have to either, I guess.  It feels more correct with it, but if it’s ignored anyway, I don’t mind.

If it works and starts minimized when you use this code, then that’s what I mean.  Your shortcut test does this behind the scenes in ShellExecute, I imagine.   I used to really miss having access to the Windows code when trying to sort out stuff like this!

So if I get what you’re saying, whoever calls CreateProcess and happens to specify STARTF_USESHOWWINDOW gets that behavior from all apps now because the window manager looks and does it for you?  

Cheers,
Dave


#include <Windows.h>
#include <tchar.h>

int _tmain(int argc, _TCHAR* argv[])
{
    STARTUPINFO si;
    ZeroMemory(&si, sizeof(si));
    si.cb = sizeof(si);
    si.dwFlags = STARTF_USESHOWWINDOW;
    si.wShowWindow = SW_SHOWMINIMIZED;

    PROCESS_INFORMATION pi;
    ZeroMemory(&pi, sizeof(pi));

    TCHAR cmdLine[] = _T("tiny.exe");
    BOOL bSuccess = CreateProcess(NULL, cmdLine, NULL, NULL, FALSE, 0, NULL, NULL, &si, &pi);
    if (!bSuccess)
    {
        _tprintf(_T("CreateProcess failed (%d)\n"), GetLastError());
        return 1;
    }

    WaitForSingleObject(pi.hProcess, INFINITE);

    CloseHandle(pi.hProcess);
    CloseHandle(pi.hThread);

    return 0;
}
